signac/core/synced_collection.py

```diff
@@ -58,6 +60,17 @@ def register(cls, *args):
         for _cls in args:
             cls.registry[_cls.backend].append(_cls)
 
+    @classmethod
+    def add_validator(cls, *args):
```

Comment:
Since you chose the classmethod approach, we need to consider how to handle the inheritance of validators. For example, have you tested whether a `JSONDict` will throw an error if you try to use a dotted key? With the current implementation I don't think that it will, because those validators only belong to the parent class but won't be inherited by the child.

Reply:
Yes, I have tested it. We also have a test for that. https://github.com/glotzerlab/signac/blob/1906db8dc0f91c3950de06ad9e370c46275b07e9/tests/test_synced_collection.py#L344-L346

Reply:
Ah OK I see the issue with the current design. The `_validators` list is defined for the `SyncedCollection` class, so when any subclass (for instance, `JSONCollection`) calls `add_validator`, it's actually just extending the same list. As a result, adding a validator to any of the classes adds it to all of them. This example illustrates the problem:

```
class A:
    _validators = list()

    @classmethod
    def add(cls, *val):
        cls._validators.extend(val)


class B(A):
    pass


class C(B):
    pass


A.add(1)
B.add(2)
C.add(3)

print(A._validators)
print(B._validators)
print(C._validators)

print(A._validators is B._validators)
print(A._validators is C._validators)

# Output 
[1, 2, 3]
[1, 2, 3]
[1, 2, 3]
True
True
```

The easiest solution that maintains the class-based approach to validators and still ensures that all instances of `SyncedCollection` get their own `_validators` list is using the `__init_subclass__` hook:

```
class A:
    @classmethod
    def __init_subclass__(cls):
        cls._validators = list()

    @classmethod
    def add(cls, *val):
        cls._validators.extend(val)


class B(A):
    pass


class C(B):
    pass


print(hasattr(A, '_validators'))
print(hasattr(B, '_validators'))
print(hasattr(C, '_validators'))
print(B._validators is C._validators)

# Output 
False
True
True
False
```

Note that this method is called whenever subclasses are created, but not for the creation of the original class, so you will still need to define the `_validators` variable in the `SyncedCollection` class. *Please document why we are using this hook when you write this method.*

Reply:
However, this reintroduces my original question in [this comment](https://github.com/glotzerlab/signac/pull/378#discussion_r475798622): if we do the class-based approach, we need to handle inheritance of validators from parent classes. @bdice I noticed your response here, what are your thoughts?  I think it would be very confusing to register a `JSONFormatValidator` to the `JSONCollection`, and then have to re-register it to the `JSONDict` and `JSONList` classes. On the other hand, as a user of this code, how would you expect this to behave if you instantiated your own synced collections? Would you expect that modifying a parent of a class changed the validators of that class? I think that is the most consistent behavior we could define.

Also, from an implementation point of view, the only way I can think of to enable this is to define `SyncedCollection.validators` as a property that walks the MRO tree and finds the validators registered to all parents. Is there a simpler alternative?

Reply:
@vyasr I read the comments above and in the linked thread. I like the class-based approach you suggested. I would like to register validators with `JSONCollection` and have those apply automatically to `JSONList` and `JSONDict`.

I haven't tried this myself so I don't know the pros/cons of the functionality of `__init_subclass__` (which I wasn't aware of) vs. walking the MRO entries. (Also in my understanding, it's not really an "MRO tree" due to C3 linearization, but that's subtle.)

Reply:
The two aren't mutually exclusive, but rather one implies the other. If validators are stored per class, then you _have_ to traverse the MRO (you're right, it's not a tree with new-style classes but I recycled old terminology there) to find the validators added to parent classes. The only way I can see to avoid this is to either 1) not share validators with subclasses, which I don't think makes sense, or 2) to use the instance-based where validators are added to every instance of a class. Option 2 is simpler from an implementation perspective than the class-based approach, but I think it's a poorer design since validation really should apply to all instances of the class, so I would prefer the class approach. What I'm asking is whether you can think of any way to avoid having to traverse the list of parent classes to find this; if not, then I think we just need to implement a `SyncedCollection.validators` `classmethod` property that collects this in the same we handle operations in the FlowProject in signac-flow.

Reply:
@vyasr I don't think we can use `classmethod` as property. I have implemented it in `_validate`. https://github.com/glotzerlab/signac/blob/39d630d5032bed09d7d08b412e726ee5067a4af8/signac/core/synced_collection.py#L160-L168

Reply:
@vishav1771 I agree with your statement that `@property` and `@classmethod` don't work together. It's possible to create a custom `@classproperty` decorator but that's too complicated in my opinion. This is an alternative using `@property` with `type(self)` instead of `cls`:

```python
class AwesomeClass:
    @property
    def name(self):
        return type(self).__name__

awesome = AwesomeClass()
print(awesome.name)
```